pkg/ruler/manager.go

```diff
@@ -391,6 +385,23 @@ func (r *DefaultMultiTenantManager) Stop() {
 	r.userManagerMtx.Unlock()
 	level.Info(r.logger).Log("msg", "all user managers stopped")
 
+	// Stop notifiers after all rule evaluations have finished, so that we have
+	// a chance to send any notifications generated while shutting down.
+	// rulerNotifier.stop() may take some time to complete if notifications need to be drained from the queue.
+	level.Info(r.logger).Log("msg", "stopping user notifiers")
```

Comment:
Is this change testable? For example, a custom manager that push a notification when stopping and then ensuring the notification is delivered?

Reply:
Yep, good idea.

I've added a test in 63af277. This will fail until https://github.com/prometheus/prometheus/pull/14290 is merged and then vendored into Mimir.